### ISSUE TITLE
Security Issue:HTTP Smuggling Detection

### ISSUE THREAD

**Issue by kbryant31415** [eclipse/jetty.project#5420]
**Jetty version**
9.4.28.v20200408
**Java version**
Any Java version compatible with jetty.
**OS type/version**
Applicable to all OS.
**Description**
Nessus has updated its rules recently. And a new security issue was found.
Posting the vulnerability information here for reference.
+++++++++++++++++++++
https://www.tenable.com/plugins/nessus/140735

HTTP Smuggling Detection
MEDIUM Nessus Plugin ID 140735
Synopsis
The remote web server is a potential candidate for HTTP Smuggling.
Description
According to the result of a remote check, the target web Server does not have some of the common mitigations in place to prevent HTTP Smuggling attacks. If HTTP Smuggling is possible a remote unauthenticated attacker could exploit this to gain access to backend resources that they would otherwise not have access to.
Solution
Refer to your vendor documentation to address this issue.
See Also
http://www.nessus.org/u?d6c4384f

Plugin Details
Severity: Medium

ID: 140735

File Name: http_smuggling_detect.nbin

Version: 1.1

Type: remote

Family: Web Servers

Published: 2020/09/22

Updated: 2020/09/22

Dependencies: 10107

Configuration: Enable paranoid mode

Risk Information
Risk Factor: Medium

CVSS Score Source: manual

CVSS Score Rationale: Based on analysis of vulnerability

CVSS v2.0
Base Score: 4

Vector: CVSS2#AV:N/AC:L/Au:S/C:N/I:P/A:N

CVSS v3.0
Base Score: 6.5

Vector: CVSS:3.0/AV:N/AC:L/PR:L/UI:N/S:U/C:N/I:H/A:N

Vulnerability Information
Required KB Items: Settings/ParanoidReport
+++++++++++++++++++++

**For this issue, the plugin output is as below:**

A web server was discovered on port 21180

  The below information could make the Web server a potential candidate for HTTP Smuggling

  - HTTP/1.1 is supported.

**Comment by joakime** [comment]
According to https://portswigger.net/research/http-desync-attacks-request-smuggling-reborn
this is all about messing around with `Content-Encoding` and `Transfer-Encoding` headers.

We have mitigation in place for most of these listed scenarios, and tests to prove it out.

Some examples ...

```
POST / HTTP/1.1
Host: example.com
Content-Length: 6
Content-Length: 5

12345GPOST / HTTP/1.1
Host: example.com
```

This one is rejected by Jetty for having 2 Content-Length headers.

and

```
POST / HTTP/1.1
Host: example.com
Content-Length: 6
Transfer-Encoding: chunked

0

GPOST / HTTP/1.1
Host: example.com
```

and

```
POST /about HTTP/1.1
Host: example.com
Transfer-Encoding: chunked
Content-Length: 4

1
Z
Q
```

and

```
POST / HTTP/1.1
Host: example.com
Content-Length: 3
Transfer-Encoding: chunked

6
PREFIX
0

POST / HTTP/1.1
Host: example.com
```

and

```
POST /about HTTP/1.1
Host: example.com
Transfer-Encoding: chunked
Content-Length: 6

0

X
```

These are rejected by Jetty for having both a Content-Encoding and a Transfer-Encoding header.

```
Transfer-Encoding:[tab]chunked
```

This is allowed (tab after the field delim), and doesn't break anything, it's just whitespace.

```
Transfer-Encoding : chunked
```

This is also allowed (whitespace before field delim), and does not break anything.

```
GET / HTTP/1.1
 Transfer-Encoding: chunked
```

This is also allowed (whitespace before field name), and does not break anything.

There are 2 variations of badly following the RFC7230 spec.

```
X: X[\n]Transfer-Encoding: chunked
```

This one doesn't mean anything (to me).
What does the "[\n]" represent here?

```
Transfer-Encoding
 : chunked
```

This one is rejected by Jetty for not having the field delim on the first line.


What I don't see, are tests for the invalid / misleading values on `Transfer-Encoding` though.

Some examples ...

```
Transfer-Encoding: xchunked
```

```
Transfer-Encoding: chunked
Transfer-Encoding: x
```

Shown as example 

```
C> POST /search HTTP/1.1
C> Host: example.com
C> Content-Type: application/x-www-form-urlencoded
C> Content-Length: 53
C> Transfer-Encoding: zchunked
C> 
C> 11
C> =x&q=smuggling&x=
C> 0
C> 
C> POST /search HTTP/1.1
C> Host: example.com
S< GET /404 HTTP/1.1
S< Foo: b
```

The specific bad Transfer-Encoding value of `zchunked` in this example is irrelevant, as the entire request is rejected for having both `Content-Encoding` and `Transfer-Encoding`.

There's no demonstrable impact of a bad Transfer-Encoding value that doesn't ALSO include a Content-Encoding header, a bad/invalid Transfer-Encoding header on it's own doesn't show a security concern on that page.

In conclusion, your security scanning tool is either reporting bogus faults/failures/issues, or you've intentionally configured Jetty to allow these concepts (we allow that in the HttpConfiguration and HttpCompliance settings), or it's being so vague and unclear that the specific fault it's concerned about isn't what you've linked to.

Get us more details, specific examples that are at fault, or cause issues, and we'll evaluate those specific concerns.
Right now we have security mitigation in place for each of these.

**Comment by sbordet** [comment]
@kbryant31415 also, this is not the way to report alleged security issues.

Please read carefully https://github.com/eclipse/jetty.project/blob/jetty-10.0.x/CONTRIBUTING.md#reporting-security-issues.

**Comment by gregw** [comment]
I'm closing this a false positive from the testing tool.  
If there are other security concerns, please follow the security reporting guidelines posted above.
